Updates from testing

alerman opened the pull request.

alerman committed:
make shaded, and add some deps

alerman committed:
Sent content type header and make shaded jar

ddanielr commented on datawave-micrometer/pom.xml:
```diff
@@ -12,8 +12,8 @@
     <artifactId>datawave-micrometer</artifactId>
 
     <properties>
-        <maven.compiler.source>17</maven.compiler.source>
```
Can we keep it as `17` and just change the compiler target to `11`? 
That would allow it to match the current accumulo build process and future desired datawave build process.

alerman committed:
update only source to 11

ddanielr commented on datawave-micrometer/pom.xml:
```diff
@@ -12,8 +12,8 @@
-        <maven.compiler.source>17</maven.compiler.source>
-        <maven.compiler.target>17</maven.compiler.target>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
```
```suggestion
        <maven.compiler.target>11</maven.compiler.target>
        <maven.compiler.release>11</maven.compiler.release>
        <minimalJavaBuildVersion>17</minimalJavaBuildVersion>
```

ctubbsii commented on datawave-micrometer/pom.xml:
```diff
@@ -12,7 +12,7 @@
 
     <properties>
-        <maven.compiler.source>17</maven.compiler.source>
+        <maven.compiler.source>11</maven.compiler.source>
         <maven.compiler.target>17</maven.compiler.target>
```
You should probably do:

```suggestion
        <maven.compiler.release>11</maven.compiler.release>
        <maven.compiler.source>${maven.compiler.release}</maven.compiler.source>
        <maven.compiler.target>${maven.compiler.release}</maven.compiler.target>
```

ctubbsii commented on datawave-micrometer/src/main/java/datawave/prometheus/DatawavePrometheusMeterRegistry.java:
```diff
@@ -62,7 +67,7 @@ public MeterRegistry create(MeterRegistryFactory.InitParameters params) {
 
             new Thread(server::start).start();
-        } catch (IOException e) {
+        } catch (Throwable e) {
```
Catching Throwable is very bad practice. It catches Error types as well, which by definition are `serious problems that a reasonable application should not try to catch.`

If you want to include RTEs along with the checked exception, it'd be better to do:

```java
        } catch (IOException | RuntimeException e) {
```

But in this case, you're just rewrapping and throwing the RTE, so it doesn't make sense to catch the RTE here.

It also doesn't make sense to downgrade an Error to a RTE.

If there are additional checked exceptions, though, you could do this multi-catch solution, though.

ddanielr replied:
Just adding context.
The original reason for the `Throwable` catch statement was because the underlying code was throwing a "ClassNotFound" Exception but not putting that in the logs or killing the process. 
Instead, the code would still host the http endpoints, requests would hit the endpoint but return no body content.

ctubbsii replied:
If it's `ClassNotFoundException`, that's a checked exception. Usually those can be handled with a catch of the parent class, `ReflectiveOperationException`. If it's `NoClassDefFoundError`, that's not generally safe to recover from, and should not be handled (except as part of an effort to kill the process). That can be caused by a variety of things that can leave the JVM in an indeterministic and unpredictable state. For example, it can happen due to an exception in a static initializer that leaves the JVM in an unpredictable state, having partially initialized some static stuff that sticks around in the JVM and corrupts other threads.

So, I'd still avoid catching Throwable. It's pretty bad practice, and can lead to suppressing more severe errors like data corruption down the line. It's better to catch the checked and runtime exceptions only.

ctubbsii commented on datawave-micrometer/src/main/java/datawave/prometheus/DatawavePrometheusMeterRegistry.java:
```diff
@@ -62,7 +67,7 @@ public MeterRegistry create(MeterRegistryFactory.InitParameters params) {
             new Thread(server::start).start();
-        } catch (IOException e) {
+        } catch (Throwable e) {
             throw new RuntimeException(e);
```
It's better to throw a more specific RTE whenever possible. In this case, it's an IOException, so using UncheckedIOException makes sense:

```suggestion
            throw new UncheckedIOException(e);
```

ddanielr committed:
Simplify poms and add .gitignore file

Moved property and version information to parent pom.
Marked accumulo-core as provided dep. Reducing shaded jar contents
drastically.

Added .gitignore file
Setup project to build with java 17 but target and release for java 11.

ddanielr committed:
Cleanup build warnings

Remove uneeded import statement
Cleanup HashMap creation
Fix log message

ddanielr committed:
Replace deprecated class usage

ddanielr committed:
Reference existing property values

Remove javadoc property as checkstyle plugin has not been added yet

ddanielr committed:
additional pom plugin fixes

ddanielr committed:
cleanup checkstyle warnings

ddanielr committed:
Cleanup additional build warnings

ddanielr committed:
Remove last build warning

alerman committed:
Update datawave-micrometer/src/main/java/datawave/prometheus/DatawavePrometheusMeterRegistry.java

Co-authored-by: Christopher Tubbs <ctubbsii@apache.org>
